Openstack Should Not Return Provisioning Error When No Internal Networks and No Network Config

manadart opened the pull request:
## Description of change

A prior patch, https://github.com/juju/juju/pull/8937 causes Openstack to return an error when there are no reported internal networks, and when there is no model configuration for network.

This change ensures that when there is no network configured, an error is only returned when multiple internal networks are reported by Openstack.

## QA steps

- Additional unit testing.
- Verified successful bootstrap on to Canonistack.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1785564

manadart commented:
@babbageclunk Can you run this against the Canonistack installation to verify it as part of the review?

manadart commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Openstack Error When No Networks Listed](https://trello.com/c/5cQYkaOy/437-openstack-error-when-no-networks-listed)

babbageclunk commented:
Yeah, will do - thanks for the quick fix!

On Mon., 6 Aug. 2018, 20:46 Joseph Phillips, <notifications@github.com>
wrote:

>
> <https://camo.githubusercontent.com/193766a3b9959c5f4ed5cd8cf3251d015e839d51/68747470733a2f2f6769746875622e7472656c6c6f2e73657276696365732f696d616765732f6d696e692d7472656c6c6f2d69636f6e2e706e67> Openstack
> Error When No Networks Listed
> <https://trello.com/c/5cQYkaOy/437-openstack-error-when-no-networks-listed>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/juju/juju/pull/9021#issuecomment-410634699>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAiYx4Shh4A27YkUfdigJjYl6bp-_kxAks5uOAJ1gaJpZM4Vv_CG>
> .
>

jameinel commented on provider/openstack/provider.go:
```diff
@@ -1001,9 +1001,16 @@ func (e *Environ) StartInstance(ctx context.ProviderCallContext, args environs.S
+			// error reports multiple Openstack networks.
+			// If there are no Openstack networks at all (such as Canonistack),
+			// having no network config is not an error condition.
+			if strings.HasPrefix(err.Error(), "multiple networks") {
```
Is there any way to have a better analysis of the error than checking the string starts with "multiple networks"?
I like the idea of this, but where does that error originate from?
How stable is that string going to be? Is there any chance there is an error code we can recognize/have Goose translate an obvious error into an error code?

manadart replied:
In this case we control that error via `processResolveNetworkIds` (in _networking.go_). I considered a custom error type, but since it's in our code, I figured having coverage keeps us safe.

hmlanigan commented on provider/openstack/local_test.go:
```diff
@@ -2680,6 +2680,22 @@ func (s *noNeutronSuite) TestUpdateGroupControllerNoNeutron(c *gc.C) {
 	))
 }
 
+func (s *noNeutronSuite) TestStartInstanceNetworkNotSetNoError(c *gc.C) {
```
Eventually, we'd love to get rid of nova networking.  Only 1 or 2, I hope, sites still use it.  It was deprecated a couple of years ago.   Once that happens, we'll loose this test.  To get something like it working with the openstack testservice, we'd need to add the equivalent of s.Nova.SetAvailabilityZones(),  s.Neutron.SetNetworks() perhaps.  At that point SetExternalNetworks might be helpful as well.

manadart commented:
@babbageclunk I got up on Canonistack and fixed that issue.
@hmlanigan I replaced the test in the _noNeutronSuite_. Can you look at the last commit and approve/iterate as appropriate?

hmlanigan commented on provider/openstack/local_test.go:
```diff
@@ -580,6 +581,27 @@ func (s *localServerSuite) TestStartInstanceNetworkNotSetReturnsError(c *gc.C) {
+	for _, net := range model.AllNetworks() {
+		model.RemoveNetwork(net.Id)
+	}
+	s.srv.Openstack.Neutron.AddNeutronModel(model)
```
Nice find!  I didn't think of using the NeutronModel functionality that way, might be useful other places too.

hmlanigan commented on provider/openstack/local_test.go:
```diff
@@ -580,6 +581,27 @@ func (s *localServerSuite) TestStartInstanceNetworkNotSetReturnsError(c *gc.C) {
 }
 
+func (s *localServerSuite) TestStartInstanceNoNetworksNetworkNotSetNoError(c *gc.C) {
+	// Modify the server that is created by default, to clear the networks.
```
Nit: s/server/openstack service/

manadart commented:
$$merge$$
